File: 프로그래머스/2/42626. 더 맵게/더 맵게.js
class MinHeap {
  constructor() {
    this.heap = [];
  }

  size() {
    return this.heap.length;
  }
    
  peek() {
    return this.heap[0];
  }
      
  push(value) {
    this.heap.push(value);
    let currentIndex = this.heap.length - 1;
    // let parentIndex = Math.floor((currentIndex - 1) / 2)

    while (
      currentIndex > 0 &&
      this.heap[currentIndex] < this.heap[Math.floor((currentIndex - 1) / 2)]
    ) {
      [this.heap[currentIndex], this.heap[Math.floor((currentIndex - 1) / 2)]] = [this.heap[Math.floor((currentIndex - 1) / 2)], this.heap[currentIndex]];
      currentIndex = Math.floor((currentIndex - 1) / 2);
    }
  }

  pop() {
    if (this.heap.length === 0) return null;
    if (this.heap.length === 1) return this.heap.pop();

    const minValue = this.heap[0];
    this.heap[0] = this.heap.pop();
    let currentIndex = 0;

    while (currentIndex * 2 + 1 < this.heap.length) {
      let minChildIndex = currentIndex * 2 + 2 < this.heap.length && this.heap[currentIndex * 2 + 2] < this.heap[currentIndex * 2 + 1] ? currentIndex * 2 + 2 : currentIndex * 2 + 1;

      if (this.heap[currentIndex] < this.heap[minChildIndex]) {
        break;
      }

      [this.heap[currentIndex], this.heap[minChildIndex]] = [this.heap[minChildIndex], this.heap[currentIndex]];
      currentIndex = minChildIndex;
    }

    return minValue;
  }
}

function solution(scoville, K) {
  const minHeap = new MinHeap();

  for (const sco of scoville) {
    minHeap.push(sco);
  }

  let mixedCount = 0;

  while (minHeap.size() >= 2 && minHeap.peek() < K) {
    const first = minHeap.pop();
    const second = minHeap.pop();
    const mixedScov = first + second * 2;
    minHeap.push(mixedScov);
    mixedCount++;
  }

  return minHeap.peek() >= K ? mixedCount : -1;
}

File: 프로그래머스/2/42626. 더 맵게/더 맵게.py
# 배열을 내림차순으로 정리한 풀이 -> 시간초과
# def solution(scoville, K):
#     count = 0
#     scoville.sort(reverse = True)
    
#     while scoville[-1] < K:
#         if len(scoville) < 2:
#             return -1
        
#         first = scoville.pop()
#         second = scoville.pop()
        
#         new_scoville = first + second * 2
#         scoville.append(new_scoville)
#         count += 1
#         scoville.sort(reverse = True)
        
#     return count

# 최소 힙 사용 풀이
import heapq

def solution(scoville, K):
    count = 0
    heapq.heapify(scoville) # scoville을 최소 힙으로 변환
    
    while scoville[0] < K:
        if len(scoville) < 2:
            return -1
        
        first = heapq.heappop(scoville)
        second = heapq.heappop(scoville)
        
        new_scoville = first + second * 2
        heapq.heappush(scoville, new_scoville)
        count += 1
        
    return count